src/lib/grub.rs

```diff
@@ -65,27 +65,57 @@ pub fn unset_boot_counter(grub_path: &str, mount_info_path: &str) -> Result<()>
 }
 
 fn unset_grub_var(key: &str, grub_path: &str, mount_info_path: &str) -> Result<()> {
-    remount_boot_rw(Path::new(mount_info_path)).context("Failed to remount /boot as rw")?;
-    Command::new("grub2-editenv")
+    let was_rw = is_boot_rw(Path::new(mount_info_path))
+        .map_err(|e| anyhow::anyhow!("Failed to check boot mount state: {}", e))?;
+
+    if !was_rw {
+        remount_boot_rw(Path::new(mount_info_path)).context("Failed to remount /boot as rw")?;
+    }
+
+    // Execute GRUB command and capture result
+    let grub_result = Command::new("grub2-editenv")
         .arg(grub_path)
         .arg("unset")
         .arg(key)
         .status()
-        .context("Unable to clear boot_counter")?;
+        .context("Unable to clear boot_counter");
+
+    // Always attempt to restore mount state, regardless of GRUB command result
+    if !was_rw {
```

Comment:
I'm a little confused here, probably through my own misunderstanding. So we check if boot is rw in line 68, then if boot is not rw we remount boot as rw in line 72. Then in line 84 we do the same check on `was_rw`, but now we remount as ro? 

I think this is probably my own lack of grub understanding, but why do we remount boot twice on the same `was_rw` output?

Reply:
I see below that we do a similar thing, so I'm sure it's just me missing something. Would love a mini explanation just for my own knowledge if possible

Reply:
the overall idea is to restore the /boot state(ro,rw) .
rpm-ostree -> rw , while bootc -> ro, its probably due to composefs.
we cannot modify grub variable if the boot is ro,so we `remount rw -> modify grub ->remount ro`- which greenboot is doing now. This is not okay for bootc but not rpm-ostree